### ISSUE TITLE
Question regarding RNN-MDN

### ISSUE THREAD

**Issue by vipinpillai** [worldmodels/worldmodels.github.io#6]
Thanks for sharing the demo. I had some questions regarding the RNN-MDN module used to sample z<sub>t+1</sub>. 

Since the latent sample is 32-dimensional for the car racing experiment, does the MDN model each distribution within the mixture as a Multivariate Normal distribution? 

Could you please share some details regarding the explicit separation between the RNN and the MDN modules and the necessary tweaks needed for stabilizing the training of the MDN and to keep the NLL loss calculation tractable for the multivariate MDN.

**Comment by hardmaru** [comment]
Hi @vipinpillai 

Please also refer to the RNN Description in Appendix:

*For the M Model, we use an LSTM  recurrent neural network combined with a Mixture Density Network as the output layer. We use this network to model the probability distribution of the next z in the next time step as a Mixture of Gaussian distribution (NB: This answers @vipinpillai's 1st question). This approach is very similar to Graves’ Generating Sequences with RNNs in the Unconditional Handwriting Generation section and also the decoder-only section of SketchRNN. The only difference in the approach used is that we did not model the correlation parameter between each element of z), and instead had the MDN-RNN output a diagonal covariance matrix of a factored Gaussian distribution. (This answers part of @vipinpillai's 2nd question regarding making training stable and loss calculation tractable)*

*Unlike the handwriting and sketch generation works, rather than using the MDN-RNN to model the pdf of the next pen stroke, we model instead the pdf of the next latent vector z. The MDN-RNNs were trained for 20 epochs on the data collected from a random policy agent. In the Car Racing task, the LSTM used 256 hidden units, while the Doom task used 512 hidden units. In both tasks, we used 5 Gaussian mixtures and did not model the correlation ρ parameter, hence z is sampled from a factored mixture of Gaussian distribution. (@vipinpillai: please keep this sampling in mind, since it helps safeguard against overfitting to a stored set of z's)*

*When training the MDN-RNN using teacher forcing from the recorded data, we store a pre-computed set of μ and σ for each of the frames, and sample an input z ∼N(μ,σ) each time we construct a training batch, to prevent overfitting our MDN-RNN to a specific sampled z.*

I put some pointers to your questions referencing the article. Please let me know if you have any more questions!

**Comment by vipinpillai** [comment]
Thanks a lot @hardmaru for providing detailed pointers. 

One last implementation specific question is whether you have considered each dimension of latent z as a mixture of Univariate Gaussians or the entire z as a mixture of Multivariate Gaussians while training the MDN-RNN. I am asking this due to the numerical issues of computing the NLL for GMM where each Gaussian is 32 dimensional. 
In the approach used by Graves’ Generating Sequences with RNNs, we have a mixture of bi-variate gaussians, and hence it is less prone to numerical instability during training.

**Comment by hardmaru** [comment]
Hi @vipinpillai,

*z* is modelled as a factored Gaussian, so the correlation ρ parameter is assumed to be zero, unlike Graves (2013) that modelled in a correlation term. This decreases the model complexity, and also makes the training much more numerically stable, as you can operate in log-space and avoid having to take the logarithm of exponentials for your computational graph.

**Comment by vipinpillai** [comment]
Thanks @hardmaru . I understand that you have not included the correlation parameter.
However, I'm still trying to understand the answer to the following question:
Have you considered each dimension of latent z as a mixture of **Univariate Gaussians** or the entire z as a mixture of **Multivariate Gaussians**?

**Comment by hardmaru** [comment]
No, I have not tried that. Feel free to try it out yourself if you think it is interesting.

**Comment by hardmaru** [comment]
hi @dariocazzani,

Good question! I also had to think about this in the experiments. Another person who reproduced the CarRacing task used a method to encourage more diversity:

https://medium.com/applied-data-science/how-to-build-your-own-world-model-using-python-and-keras-64fb388ba459

Have you thought about potential approaches that might help overcome this issue? I'd be curious to know what other people come up with! Feel free to list some here.

I'll let you know what I did to generate the random actions to encourage a more diverse set, and I think my approach is more elegant than simply hitting the accelerate pedal :)

**Comment by dariocazzani** [comment]
The person who wrote that blog post did a few things: 
- Limited the length of the rollouts to 300 
- The policy for generating actions is straighforward (prefer keeping the previous action 1/5 of the times, some other tricks) and it starts by going straight for 60 steps for every single rollout.

This works well if the car starts at the same spot and if we limit the rollouts at 300.

What I did was first of all to make the **car start at random points** for each rollout:
```python
from gym.envs.box2d.car_dynamics import Car
from gym.envs.box2d import CarRacing
[...]
position = np.random.randint(len(env.track))        
env.car = Car(env.world, *env.track[position][1:4])
```
I made a **PR** to OpenAI gym to make this possible without "tricks" [Link to PR](https://github.com/openai/gym/pull/1025)

And this is my policy for generating random actions.
Notice 2 things:
- I can not count on a straight road at the beginning
- I am assuming that the car never has to brake and accelerate at the same time. Thus I could **reduce the number of actions to 2**. This was beneficial for reducing by ~33% the number of parameters of the **Controller** 
```python
def generate_action(prev_action):
    if np.random.randint(3) % 3:
        return prev_action

    index = np.random.randn(3)
    # Favor acceleration over the others:
    index[1] = np.abs(index[1])
    index = np.argmax(index)
    mask = np.zeros(3)
    mask[index] = 1

    action = np.random.randn(3)
    action = np.tanh(action)
    action[1] = (action[1] + 1) / 2
    action[2] = (action[2] + 1) / 2

return action*mask
```

I am documenting everything on a series of blog posts (still WIP since I need to work on it on my spare time).
The first episode is [World Models in TensorFlow — Episode 1.0 — OpenAi Gym Race Car](https://medium.com/@dariocazzani/world-models-in-tensorflow-episode-1-2b3c217ebc8f)
Counting on releasing the work on **Doom** soon :)

Suggestions for improvements are always welcome :)

Thanks again for the feedback @hardmaru

**Comment by dariocazzani** [comment]
Hey @hardmaru, I'll respond point per point.
* I changed the starting point for training and data generation in order to improve robustness (idea inspired by one of your articles).
Good point to keep in mind not to change it for evaluations that can be compared with known results.
* Dimensionality reduction FTW 👍 
* Very smart idea, I'll give it a try.

**An extra bit: Variation over using MDNs for M** 
I was thinking to experiment with the RNN outputting 2 vectors, mean and std of the predicted frame, and compute the cost as the KL divergence between these 2 vectors and the VAE outputs.
Does it make sense? Have you tried it yet?
I'll see how it goes.
